### DISCUSSION TITLE
python-miio on windows 10

### DISCUSSION THREAD

**Discussion by robs23** [rytilahti/python-miio#951] [Q&A] [answered]
I struggle to use your miiocli tool on windows 10. I have python 3.8.5 installed. I installed python-miio through pip and see your libraries have been added to python's lib folder. However, when trying to run "miiocli --help" in windows cmd I'm getting "'miiocli' is not recognized". When running it from python's IDLE I'm getting "NameError: name 'miiocli' is not recognized". What else can I do? Or maybe I'm doing something wrong?

**Comment by rytilahti** [comment, marked as answer]
So, I don't have any idea about how Windows works in this case, but quick browsing suggests that you have to add `Scripts` folder to your path to make use of installed scripts (like `miiocli`). https://docs.python.org/3/using/windows.html#finding-the-python-executable suggests that the python installer offers doing this for you, do you see the correct path in your `%PATH%`? The `miiocli` should be installed inside that `Scripts` folder.

**Reply by robs23** [reply]
Thanks for prompt answer. I see there is a `Scripts` folder under my python installation and that it is a part of `%PATH%` but there's no miiocli in this `Scripts` folder (I assume it should be there?). In this folder there's `pip.exe` and I can call it from cmd and it works, so I suppose it's not my path.  So, if I'm assuming correctly, there should be `miiocli.exe` in `Scripts` folder, right?

**Reply by rytilahti** [reply]
Hi, sorry for the late reply. Are you still having this problem? Could it be that the package was not properly installed due to a missing compiler?

I just gave this a try on Windows 10 (first time using python and pip on windows) with python installed using the installer from python.org (version 3.9.2), and I can see the `exe` files for the entry points (e.g., `miio.exe`, `mirobo.exe`) alongside pip binary under `%APPDATA%\Local\Programs\Python\Python39\Scripts`.

I had to use `--no-deps` for `pip.exe` to avoid installing the VC++ to compile the library required by one pypi package required by the firmware updater (`netifaces`, I think this should be made optional to simplify the installation, see #964).

**Reply by robs23** [reply]
Hi. Well, I think it's possible. I was installing either pip or your packages when I got "error: Microsoft Visual C++ 14.0 is required" error. I then installed VC++ but it's possible that the packages were not properly installed. Having checked and I don't have those exes under the path.

**Reply by rytilahti** [reply]
Did you try installing without the dependencies? That should work and worked fine for me as I mentioned, but you cannot do firmware updates using the tool.

**Reply by robs23** [reply]
Yes, sorry for late reply. I installed it without dependencies and it went smoothly

**Comment by martyanov1990** [comment]
Hello. When issuing "miiocli --help" an error "Traceback (most recent call last):
  File "C:\python\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\python\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\python\Scripts\miiocli.exe\__main__.py", line 4, in <module>
  File "C:\python\lib\site-packages\miio\cli.py", line 43, in <module>
    cli.add_command(device_class.get_device_group())
  File "C:\python\lib\site-packages\miio\integrations\vacuum\roborock\vacuum.py", line 932, in get_device_group
    @dg.resultcallback()
AttributeError: 'DeviceGroup' object has no attribute 'resultcallback'. Did you mean: 'result_callback'?" is returned. 
What could be wrong?

**Reply by serious-angel** [reply]
The same issue:
<img width="650" alt="NVIDIA_Share_NVIDIA_GeForce_Overlay_20220623_021719_Uieup" src="https://user-images.githubusercontent.com/25136754/175181457-82114b55-641b-4279-b33c-2fcab05c90b6.png">.
The relative solution: https://github.com/rytilahti/python-miio/issues/1389#issuecomment-1095212507
